File: scripts/nba/local/scrape-players-nba-api.js
#!/usr/bin/env node

/**
 * NBA Player Historical Data Scraper (NBA Stats API)
 * 
 * Uses official NBA Stats API instead of scraping HTML
 * Much more reliable and faster than Basketball-Reference scraping
 * 
 * Scrapes 5 seasons (2020-21 through 2024-25) of player data
 * 
 * Usage:
 *   node scripts/nba/local/scrape-players-nba-api.js
 * 
 * Runtime: ~5-10 minutes
 */

import fetch from 'node-fetch';
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Seasons to scrape
const SEASONS = [
  { label: '2020-21', api: '2020-21' },
  { label: '2021-22', api: '2021-22' },
  { label: '2022-23', api: '2022-23' },
  { label: '2023-24', api: '2023-24' },
  { label: '2024-25', api: '2024-25' }
];

// Rate limiting
const DELAY_BETWEEN_REQUESTS = 1000; // 1 second
const DELAY_BETWEEN_SEASONS = 2000;  // 2 seconds

// NBA Stats API headers (required!)
const NBA_HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
  'Accept': 'application/json, text/plain, */*',
  'Accept-Language': 'en-US,en;q=0.5',
  'Referer': 'https://www.nba.com/',
  'Origin': 'https://www.nba.com',
  'Connection': 'keep-alive',
  'x-nba-stats-origin': 'stats',
  'x-nba-stats-token': 'true'
};

/**
 * Delay helper
 */
function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

/**
 * Fetch with retry logic
 */
async function fetchWithRetry(url, retries = 3) {
  for (let i = 0; i < retries; i++) {
    try {
      const response = await fetch(url, { headers: NBA_HEADERS });
      
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }
      
      return await response.json();
    } catch (error) {
      console.error(`  ⚠️  Attempt ${i + 1}/${retries} failed: ${error.message}`);
      
      if (i < retries - 1) {
        await delay(3000); // Wait 3 seconds before retry
      } else {
        throw error;
      }
    }
  }
}

/**
 * Scrape player stats for a season
 */
async function scrapeSeasonPlayers(season) {
  console.log(`\n🏀 Scraping ${season.label}...`);
  
  // Fetch traditional stats
  const tradUrl = `https://stats.nba.com/stats/leaguedashplayerstats?` + new URLSearchParams({
    Season: season.api,
    SeasonType: 'Regular Season',
    PerMode: 'Totals',
    MeasureType: 'Base'
  });
  
  console.log(`  📥 Fetching traditional stats...`);
  const tradData = await fetchWithRetry(tradUrl);
  
  await delay(DELAY_BETWEEN_REQUESTS);
  
  // Fetch advanced stats
  const advUrl = `https://stats.nba.com/stats/leaguedashplayerstats?` + new URLSearchParams({
    Season: season.api,
    SeasonType: 'Regular Season',
    PerMode: 'Totals',
    MeasureType: 'Advanced'
  });
  
  console.log(`  📥 Fetching advanced stats...`);
  const advData = await fetchWithRetry(advUrl);
  
  // Parse traditional stats
  const tradHeaders = tradData.resultSets[0].headers;
  const tradRows = tradData.resultSets[0].rowSet;
  
  const players = {};
  
  for (const row of tradRows) {
    const playerData = {};
    tradHeaders.forEach((header, i) => {
      playerData[header] = row[i];
    });
    
    const playerId = playerData.PLAYER_ID;
    const playerName = playerData.PLAYER_NAME;
    const team = playerData.TEAM_ABBREVIATION;
    
    players[playerId] = {
      player: playerName,
      team: team,
      season: season.label,
      
      // Games
      games_played: playerData.GP || 0,
      games_started: playerData.GS || 0,
      minutes_played: playerData.MIN || 0,
      
      // Shooting
      fgm: playerData.FGM || 0,
      fga: playerData.FGA || 0,
      fg_pct: playerData.FG_PCT || 0,
      fg3m: playerData.FG3M || 0,
      fg3a: playerData.FG3A || 0,
      fg3_pct: playerData.FG3_PCT || 0,
      ftm: playerData.FTM || 0,
      fta: playerData.FTA || 0,
      ft_pct: playerData.FT_PCT || 0,
      
      // Rebounds
      oreb: playerData.OREB || 0,
      dreb: playerData.DREB || 0,
      reb: playerData.REB || 0,
      
      // Other stats
      ast: playerData.AST || 0,
      stl: playerData.STL || 0,
      blk: playerData.BLK || 0,
      tov: playerData.TOV || 0,
      pf: playerData.PF || 0,
      pts: playerData.PTS || 0,
      
      // Plus/Minus
      plus_minus: playerData.PLUS_MINUS || 0
    };
  }
  
  // Merge advanced stats
  const advHeaders = advData.resultSets[0].headers;
  const advRows = advData.resultSets[0].rowSet;
  
  let advMerged = 0;
  for (const row of advRows) {
    const playerData = {};
    advHeaders.forEach((header, i) => {
      playerData[header] = row[i];
    });
    
    const playerId = playerData.PLAYER_ID;
    
    if (players[playerId]) {
      // Advanced metrics
      players[playerId].off_rating = playerData.OFF_RATING || null;
      players[playerId].def_rating = playerData.DEF_RATING || null;
      players[playerId].net_rating = playerData.NET_RATING || null;
      players[playerId].ast_pct = playerData.AST_PCT || null;
      players[playerId].ast_ratio = playerData.AST_RATIO || null;
      players[playerId].oreb_pct = playerData.OREB_PCT || null;
      players[playerId].dreb_pct = playerData.DREB_PCT || null;
      players[playerId].reb_pct = playerData.REB_PCT || null;
      players[playerId].tov_pct = playerData.TOV_PCT || null;
      players[playerId].efg_pct = playerData.EFG_PCT || null;
      players[playerId].ts_pct = playerData.TS_PCT || null;
      players[playerId].usg_pct = playerData.USG_PCT || null;
      players[playerId].pace = playerData.PACE || null;
      players[playerId].pie = playerData.PIE || null;
      
      advMerged++;
    }
  }
  
  console.log(`  ✅ Scraped ${Object.keys(players).length} players`);
  console.log(`  ✅ Merged ${advMerged} advanced stats`);
  
  // Filter to significant players (at least 5 games or 50 minutes)
  const filteredPlayers = Object.values(players).filter(p => 
    p.games_played >= 5 || p.minutes_played >= 50
  );
  
  console.log(`  ✅ Filtered to ${filteredPlayers.length} significant players`);
  
  return filteredPlayers;
}

/**
 * Main scraper
 */
async function scrapeAllSeasons() {
  console.log('🏀 NBA Player Historical Data Scraper (NBA Stats API)');
  console.log('='.repeat(60));
  console.log(`Seasons: ${SEASONS.map(s => s.label).join(', ')}`);
  console.log('='.repeat(60));
  
  const allPlayers = [];
  
  for (let i = 0; i < SEASONS.length; i++) {
    const season = SEASONS[i];
    
    try {
      const players = await scrapeSeasonPlayers(season);
      allPlayers.push(...players);
      
      // Save individual season file
      const seasonFile = path.join(
        __dirname,
        '../../../data/nba/players/archive',
        `player_seasons_${season.label.replace('-', '_')}.json`
      );
      
      await fs.mkdir(path.dirname(seasonFile), { recursive: true });
      await fs.writeFile(seasonFile, JSON.stringify({
        schema_version: 1,
        scraped_at: new Date().toISOString(),
        season: season.label,
        source: 'nba-stats-api',
        player_count: players.length,
        players
      }, null, 2));
      
      console.log(`  💾 Saved: ${seasonFile}\n`);
      
      // Delay before next season
      if (i < SEASONS.length - 1) {
        console.log(`  ⏱️  Waiting ${DELAY_BETWEEN_SEASONS}ms...\n`);
        await delay(DELAY_BETWEEN_SEASONS);
      }
    } catch (error) {
      console.error(`  ❌ Failed to scrape ${season.label}: ${error.message}\n`);
    }
  }
  
  // Save combined archive
  const combinedFile = path.join(
    __dirname,
    '../../../data/nba/players/archive',
    'player_seasons_combined.json'
  );
  
  await fs.writeFile(combinedFile, JSON.stringify({
    schema_version: 1,
    scraped_at: new Date().toISOString(),
    seasons: SEASONS.map(s => s.label),
    source: 'nba-stats-api',
    total_player_seasons: allPlayers.length,
    players: allPlayers
  }, null, 2));
  
  console.log('\n' + '='.repeat(60));
  console.log('✅ SCRAPING COMPLETE');
  console.log('='.repeat(60));
  console.log(`📁 Combined file: ${combinedFile}`);
  console.log(`📊 Total player-seasons: ${allPlayers.length}`);
  console.log(`🏀 Average per season: ${Math.round(allPlayers.length / SEASONS.length)}`);
  console.log('\n💡 Next steps:');
  console.log('  1. Run team scraper: node scripts/nba/local/scrape-teams-nba-api.js');
  console.log('  2. Calculate RCI: node scripts/nba/local/build-rosters-with-rci.js');
  console.log('  3. Validate: node scripts/nba/local/validate-data.js');
}

// Run scraper
scrapeAllSeasons().catch(error => {
  console.error('\n❌ FATAL ERROR:', error);
  process.exit(1);
});

File: scripts/nba/local/scrape-players-nba-api.py
#!/usr/bin/env python3
"""
NBA Player Historical Data Scraper (nba_api)

Uses the nba_api library for reliable NBA Stats API access
Scrapes 5 seasons (2020-21 through 2024-25) for ALL 30 teams

Usage:
    python3 scripts/nba/local/scrape-players-nba-api.py

Runtime: ~5-10 minutes
"""

import json
import time
from pathlib import Path
from nba_api.stats.endpoints import leaguedashplayerstats
from nba_api.stats.library.parameters import SeasonTypeAllStar

# Seasons to scrape (last 5 seasons + current)
SEASONS = [
    {'label': '2021-22', 'api': '2021-22'},
    {'label': '2022-23', 'api': '2022-23'},
    {'label': '2023-24', 'api': '2023-24'},
    {'label': '2024-25', 'api': '2024-25'},
    {'label': '2025-26', 'api': '2025-26'}
]

# Rate limiting
DELAY_BETWEEN_REQUESTS = 1.5  # seconds
DELAY_BETWEEN_SEASONS = 3.0   # seconds

def scrape_season_players(season):
    """Scrape player stats for a single season"""
    print(f"\n🏀 Scraping {season['label']}...")
    
    try:
        # Fetch traditional stats
        print(f"  📥 Fetching traditional stats...")
        trad_stats = leaguedashplayerstats.LeagueDashPlayerStats(
            season=season['api'],
            season_type_all_star=SeasonTypeAllStar.regular,
            per_mode_detailed='Totals',
            measure_type_detailed_defense='Base'
        )
        trad_df = trad_stats.get_data_frames()[0]
        
        time.sleep(DELAY_BETWEEN_REQUESTS)
        
        # Fetch advanced stats
        print(f"  📥 Fetching advanced stats...")
        adv_stats = leaguedashplayerstats.LeagueDashPlayerStats(
            season=season['api'],
            season_type_all_star=SeasonTypeAllStar.regular,
            per_mode_detailed='Totals',
            measure_type_detailed_defense='Advanced'
        )
        adv_df = adv_stats.get_data_frames()[0]
        
        # Merge dataframes on PLAYER_ID
        merged_df = trad_df.merge(
            adv_df,
            on='PLAYER_ID',
            suffixes=('', '_adv')
        )
        
        # Convert to our schema
        players = []
        for _, row in merged_df.iterrows():
            player = {
                'player': row['PLAYER_NAME'],
                'team': row['TEAM_ABBREVIATION'],
                'season': season['label'],
                
                # Games
                'games_played': int(row['GP']) if row['GP'] else 0,
                'games_started': int(row.get('GS', 0)) if row.get('GS') else 0,
                'minutes_played': float(row['MIN']) if row['MIN'] else 0.0,
                
                # Shooting
                'fgm': float(row['FGM']) if row['FGM'] else 0.0,
                'fga': float(row['FGA']) if row['FGA'] else 0.0,
                'fg_pct': float(row['FG_PCT']) if row['FG_PCT'] else 0.0,
                'fg3m': float(row['FG3M']) if row['FG3M'] else 0.0,
                'fg3a': float(row['FG3A']) if row['FG3A'] else 0.0,
                'fg3_pct': float(row['FG3_PCT']) if row['FG3_PCT'] else 0.0,
                'ftm': float(row['FTM']) if row['FTM'] else 0.0,
                'fta': float(row['FTA']) if row['FTA'] else 0.0,
                'ft_pct': float(row['FT_PCT']) if row['FT_PCT'] else 0.0,
                
                # Rebounds
                'oreb': float(row['OREB']) if row['OREB'] else 0.0,
                'dreb': float(row['DREB']) if row['DREB'] else 0.0,
                'reb': float(row['REB']) if row['REB'] else 0.0,
                
                # Other stats
                'ast': float(row['AST']) if row['AST'] else 0.0,
                'stl': float(row['STL']) if row['STL'] else 0.0,
                'blk': float(row['BLK']) if row['BLK'] else 0.0,
                'tov': float(row['TOV']) if row['TOV'] else 0.0,
                'pf': float(row['PF']) if row['PF'] else 0.0,
                'pts': float(row['PTS']) if row['PTS'] else 0.0,
                'plus_minus': float(row['PLUS_MINUS']) if row['PLUS_MINUS'] else 0.0,
                
                # Advanced metrics
                'off_rating': float(row['OFF_RATING']) if row.get('OFF_RATING') else None,
                'def_rating': float(row['DEF_RATING']) if row.get('DEF_RATING') else None,
                'net_rating': float(row['NET_RATING']) if row.get('NET_RATING') else None,
                'ast_pct': float(row['AST_PCT']) if row.get('AST_PCT') else None,
                'ast_ratio': float(row['AST_RATIO']) if row.get('AST_RATIO') else None,
                'oreb_pct': float(row['OREB_PCT']) if row.get('OREB_PCT') else None,
                'dreb_pct': float(row['DREB_PCT']) if row.get('DREB_PCT') else None,
                'reb_pct': float(row['REB_PCT']) if row.get('REB_PCT') else None,
                'tov_pct': float(row.get('TM_TOV_PCT', 0)) if row.get('TM_TOV_PCT') else None,
                'efg_pct': float(row['EFG_PCT']) if row.get('EFG_PCT') else None,
                'ts_pct': float(row['TS_PCT']) if row.get('TS_PCT') else None,
                'usg_pct': float(row['USG_PCT']) if row.get('USG_PCT') else None,
                'pace': float(row.get('PACE', 0)) if row.get('PACE') else None,
                'pie': float(row.get('PIE', 0)) if row.get('PIE') else None,
                
                # Note: BPM and VORP not available in NBA Stats API
                # These would need to be calculated or scraped from Basketball-Reference
                'bpm': None,
                'vorp': None
            }
            players.append(player)
        
        # Filter to significant players
        filtered_players = [
            p for p in players 
            if p['games_played'] >= 5 or p['minutes_played'] >= 50
        ]
        
        print(f"  ✅ Scraped {len(players)} players")
        print(f"  ✅ Filtered to {len(filtered_players)} significant players")
        
        return filtered_players
        
    except Exception as e:
        print(f"  ❌ Failed to scrape {season['label']}: {e}")
        return []

def main():
    """Main scraper function"""
    print("🏀 NBA Player Historical Data Scraper (nba_api)")
    print("=" * 60)
    print(f"Seasons: {', '.join([s['label'] for s in SEASONS])}")
    print("=" * 60)
    
    all_players = []
    
    # Create output directory
    output_dir = Path(__file__).parent.parent.parent.parent / 'data' / 'nba' / 'players' / 'archive'
    output_dir.mkdir(parents=True, exist_ok=True)
    
    for i, season in enumerate(SEASONS):
        players = scrape_season_players(season)
        all_players.extend(players)
        
        # Save individual season file
        season_file = output_dir / f"player_seasons_{season['label'].replace('-', '_')}.json"
        season_data = {
            'schema_version': 1,
            'scraped_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
            'season': season['label'],
            'source': 'nba-stats-api-via-nba_api-library',
            'player_count': len(players),
            'players': players
        }
        
        with open(season_file, 'w') as f:
            json.dump(season_data, f, indent=2)
        
        print(f"  💾 Saved: {season_file}\n")
        
        # Delay before next season
        if i < len(SEASONS) - 1:
            print(f"  ⏱️  Waiting {DELAY_BETWEEN_SEASONS}s...\n")
            time.sleep(DELAY_BETWEEN_SEASONS)
    
    # Save combined archive
    combined_file = output_dir / 'player_seasons_combined.json'
    combined_data = {
        'schema_version': 1,
        'scraped_at': time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime()),
        'seasons': [s['label'] for s in SEASONS],
        'source': 'nba-stats-api-via-nba_api-library',
        'total_player_seasons': len(all_players),
        'players': all_players
    }
    
    with open(combined_file, 'w') as f:
        json.dump(combined_data, f, indent=2)
    
    print("\n" + "=" * 60)
    print("✅ SCRAPING COMPLETE")
    print("=" * 60)
    print(f"📁 Combined file: {combined_file}")
    print(f"📊 Total player-seasons: {len(all_players)}")
    print(f"🏀 Average per season: {len(all_players) // len(SEASONS)}")
    print("\n💡 Next steps:")
    print("  1. Run team scraper: python3 scripts/nba/local/scrape-teams-nba-api.py")
    print("  2. Calculate RCI: node scripts/nba/local/build-rosters-with-rci.js")
    print("  3. Validate: node scripts/nba/local/validate-data.js")

if __name__ == '__main__':
    main()
